src/content/docs/en/guides/actions.mdx

```diff
@@ -0,0 +1,501 @@
+
+- In your application code, all errors are passed to the `error` object on an action result. This avoids the need for `undefined` checks on data, and allows you to display targeted feedback to the user depending on what went wrong.
+
+#### Creating an `ActionError`
+
+To throw an error, import the `ActionError()` class from the `astro:actions` module. Pass it a human-readable status `code` (e.g. `"NOT_FOUND"` or `"BAD_REQUEST"`), and an optional `message` to provide further information about the error.
+
+This example throws an error from a `likePost` action when a user is not logged in, after checking a hypothetical "user-session" cookie for authentication:
+
+```ts title="src/actions/index.ts" ins="ActionError" ins={9-12}
+import { defineAction, ActionError } from "astro:actions";
+import { z } from "astro:schema";
+
+export const server = {
+  likePost: defineAction({
+    input: z.object({ postId: z.string() }),
+    handler: async (input, ctx) => {
+      if (!ctx.cookies.has('user-session')) {
+        throw new ActionError({
+          code: "UNAUTHORIZED",
+          message: "User must be logged in.",
+        });
+      }
+      // Otherwise, like the post
+    },
+  }),
+};
+```
+
+#### Handling an `ActionError`
+
+To handle this error, you can call the action from your application and check whether an `error` property is present. This property will be of type `ActionError`, and both your code and message will be available to access.
+
+In the following example, a `LikeButton.tsx` component calls the `likePost()` action when clicked. If an authentication error occurs, the `error.code` attribute is used to determine whether to display a login link:
+
+```tsx title=src/components/LikeButton.tsx ins="if (error.code === 'UNAUTHORIZED') setShowLogin(true);"
+import { actions } from 'astro:actions';
+import { useState } from 'preact/hooks';
+
+export function LikeButton({ postId }: { postId: string }) {
+  const [showLogin, setShowLogin] = useState(false);
+  return (
+    <>
+      {
+        showLogin && <a href="/signin">Log in to like a post.</a>
+      }
+      <button onClick={async () => {
+        const { data, error } = await actions.likePost({ postId });
+        if (error?.code === 'UNAUTHORIZED') setShowLogin(true);
+        // Early return for unexpected errors
+        else if (error) return;
+        // update likes
+      }}>
+        Like
+      </button>
+    </>
+  )
+}
+```
+
+## Handling client redirects
+
+You may need to redirect to a new page when an action succeeds. When calling actions from the client, you can integrate with a client-side library like `react-router` or you can use Astro's own client-side navigation utilities. `astro:transitions/client` exposes the `navigate()` function to redirect to a new page with a nice fade transition [when view transitions are enabled](https://docs.astro.build/en/guides/view-transitions/).
```

Comment:
```suggestion
You may need to redirect to a new page when an action succeeds. When calling actions from the client, you can integrate with a client-side library like `react-router` or you can use Astro's own client-side navigation utilities. `astro:transitions/client` exposes the `navigate()` function to redirect to a new page with a nice fade transition [when view transitions are enabled](/en/guides/view-transitions/).
```
Just correcting the absolute link here!

Also leaving myself a note for final editing polish that:

- We may want to link directly to https://docs.astro.build/en/guides/view-transitions/#trigger-navigation for a description of triggering navigation with view transitions
- Our redirects page on the routing page currently doesn't make any mention of this way of redirecting, and maybe it should?
- There's no section on the API reference page for `astro:transitions` and no proper entry for `navigate()`

BIG KETTLE OF FISH HERE, and something else for #7961 ?  :sweat_smile:

Reply:
These are good callouts! I also noticed there wasn't a mention of this on the API reference. I added a note to that PR you linked.

Reply:
Just following up: 

We have an open PR to document view transitions API including `navigate()` here: https://github.com/withastro/docs/pull/9174 - it might not be merged in time to link there from this page, but that's fine.